src/Elasticsearch.Extensions.Logging/ElasticsearchLogger.cs

```diff
@@ -0,0 +1,359 @@
+				logEvent.Trace = new Trace() { Id = activity.RootId };
+
+				// Unique identifier of the transaction.
+				// A transaction is the highest level of work measured within a service, such as a request to a server.
+				var spanId = ExtractW3cSpanIdFromActivityId(activity.Id);
+				logEvent.Transaction = new Transaction() { Id = spanId ?? activity.Id };
+			}
+			else
+			{
+				if (!System.Diagnostics.Trace.CorrelationManager.ActivityId.Equals(Guid.Empty))
+				{
+					logEvent.Trace =
+						new Trace() { Id = System.Diagnostics.Trace.CorrelationManager.ActivityId.ToString() };
+				}
+			}
+		}
+
+		private LogEvent BuildLogEvent<TState>(string categoryName, LogLevel logLevel,
+			EventId eventId, TState state, Exception? exception,
+			Func<TState, Exception, string> formatter
+		)
+		{
+			var logEvent = new LogEvent();
+
+			logEvent.Ecs = _dataProcessor.GetEcs();
+			logEvent.Timestamp = ElasticsearchLoggerProvider.LocalDateTimeProvider();
+			logEvent.Message = formatter(state, exception!);
+			logEvent.Log = new Log { Level = logLevel.ToString(), Logger = categoryName };
+			logEvent.Event =
+				new Event { Action = eventId.Name, Code = eventId.Id.ToString(), Severity = _dataProcessor.GetSeverity(logLevel) };
+
+			if (exception != null) AddException(exception, logEvent);
+
+			logEvent.Agent = _dataProcessor.GetAgent();
+			logEvent.Service = _dataProcessor.GetService();
+
+			if (Options.Tags != null && Options.Tags.Length > 0) logEvent.Tags = Options.Tags;
+
+			if (Options.IncludeHost) logEvent.Host = _dataProcessor.GetHost();
+
+			if (Options.IncludeProcess) logEvent.Process = _dataProcessor.GetProcess();
+
+			if (Options.IncludeUser)
+			{
+				logEvent.User = new User
+				{
+					Id = Thread.CurrentPrincipal?.Identity.Name, Name = Environment.UserName, Domain = Environment.UserDomainName
+				};
+			}
+
+			AddTracing(logEvent);
+
+			if (Options.IncludeScopes) AddScopeValues(logEvent);
+
+			// These will overwrite any scope values with the same name
+			AddStateValues(state, logEvent);
+
+			return logEvent;
+		}
+
+		private bool CheckTracingValues(LogEvent logEvent, KeyValuePair<string, object> kvp)
+		{
+			if (kvp.Key == "trace.id")
+			{
+				var value = FormatValue(kvp.Value);
+				if (!string.IsNullOrWhiteSpace(value))
+				{
+					if (logEvent.Trace == null) logEvent.Trace = new Trace();
+
+					logEvent.Trace.Id = value;
+				}
+
+				return true;
+			}
+
+			if (kvp.Key == "transaction.id")
+			{
+				var value = FormatValue(kvp.Value);
+				if (!string.IsNullOrWhiteSpace(value))
+				{
+					if (logEvent.Transaction == null) logEvent.Transaction = new Transaction();
+
+					logEvent.Transaction.Id = value;
+				}
+
+				return true;
+			}
+
+			return false;
+		}
+
+		private string? ExtractW3cSpanIdFromActivityId(string activityId)
+		{
+			if (activityId.Length != 2 + 32 + 16 + 2 + 3) return null;
```

Comment:
Probably an edge case, but I think a non-W3C activityId can also be `2 + 32 + 16 + 2 + 3` long. 

An idea: if it's not W3C then it's probably the hierarchical id and that always starts with `|`, so a check like this would also catch that:

```
if (activityId[0] == '|' || activityId.Length != 2 + 32 + 16 + 2 + 3) return null;
```